#if (!require("BiocManager", quietly = TRUE)) install.packages("BiocManager")
#BiocManager::install("VariantAnnotation")
# if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
#BiocManager::install("StructuralVariantAnnotation")
#install.packages("stringr")
#if (!require("BiocManager", quietly = TRUE)) install.packages("BiocManager")
#BiocManager::install("StructuralVariantAnnotation")
library(VariantAnnotation)
library(StructuralVariantAnnotation)
library(stringr)
#' Simple SV type classifier
simpleEventType <- function(gr) {
  pgr = partner(gr)
  return(ifelse(seqnames(gr) != seqnames(pgr), "CTX", # inter-chromosomosal
    ifelse(strand(gr) == strand(pgr), "INV",
      ifelse(gr$insLen >= abs(gr$svLen) * 0.7, "INS", # TODO: improve classification of complex events
        ifelse(xor(start(gr) < start(pgr), strand(gr) == "-"), "DEL",
          "DUP")))))
}

# using the example in the GRIDSS /example directory
# vcf <- readVcf(".../combination_sv/step1_refine_sv/dragen/HG002_dragen_v4.0.3_hg19_20230323.sv.vcf", "hg19")
# vcf <- readVcf(".../combination_sv/step1_refine_sv/HG002_all/HG002.novaseq.pcr-free.35x.delly.vcf", "hg19")
# vcf <- readVcf(".../combination_sv/step1_refine_sv/HG002_all/HG002.novaseq.pcr-free.35x.gridss.vcf", "hg19")
# vcf <- readVcf(".../combination_sv/step1_refine_sv/HG002_all/HG002.novaseq.pcr-free.35x.lumpy.vcf", "hg19")
# vcf <- readVcf(".../combination_sv/step1_refine_sv/HG002_all/HG002.novaseq.pcr-free.35x.manta.vcf", "hg19")
# vcf <- readVcf(".../combination_sv/step1_refine_sv/HG002_all/HG002.novaseq.pcr-free.35x.svaba.vcf", "hg19")


info(header(vcf)) = unique(as(rbind(as.data.frame(info(header(vcf))), 
                                    data.frame(row.names=c("SIMPLE_TYPE"),
                                               Number=c("1"),
                                               Type=c("String"),
                                               Description=c("Simple event type annotation based purely on breakend position and orientation.")),
                                    data.frame(row.names=c("SVLEN"),
                                               Number=c("1"),
                                               Type=c("Integer"),
                                               Description=c("SV_length"))), "DataFrame"))
gr <- breakpointRanges(vcf)
svtype <- simpleEventType(gr)
info(vcf)$SIMPLE_TYPE <- NA_character_
info(vcf[gr$sourceId])$SIMPLE_TYPE <- svtype
info(vcf)$SVLEN <- NA_character_
# info(vcf[gr$sourceId])$SVLEN <- gr$svLen #GRIDSS and others
info(vcf)$SVLEN <- info(vcf)$SPAN #svABA
# writeVcf(vcf, ".../combination_sv/step1_refine_sv/dragen/HG002.annotated.dragen.vcf")
# writeVcf(vcf, ".../combination_sv/step1_refine_sv/HG002.annotated/HG002.annotated.delly.vcf") # generated by example/gridss.sh
# writeVcf(vcf, ".../combination_sv/step1_refine_sv/HG002.annotated/HG002.annotated.gridss.vcf") 
# writeVcf(vcf, ".../combination_sv/step1_refine_sv/HG002.annotated/HG002.annotated.lumpy.vcf") 
# writeVcf(vcf, ".../combination_sv/step1_refine_sv/HG002.annotated/HG002.annotated.manta.vcf") 
# writeVcf(vcf, ".../combination_sv/step1_refine_sv/HG002.annotated/HG002.annotated.svaba.vcf") 
# TODO: perform event filtering here
# By default, GRIDSS is very sensitive but this comes at the cost of a high false discovery rate
gr <- gr[gr$FILTER == "PASS" & partner(gr)$FILTER == "PASS"] # Remove low confidence calls

simplegr <- gr[simpleEventType(gr) %in% c("INS", "INV", "DEL", "DUP")]
simplebed <- data.frame(
	chrom=seqnames(simplegr),
	# call the centre of the homology/inexact interval
	start=as.integer((start(simplegr) + end(simplegr)) / 2),
	end=as.integer((start(partner(simplegr)) + end(partner(simplegr))) / 2),
	name=simpleEventType(simplegr),
	score=simplegr$QUAL,
	strand="."
)
# Just the lower of the two breakends so we don't output everything twice
# simplebed <- simplebed[simplebed$start < simplebed$end,]
#write.table(simplebed, ".../combination_sv/step1_refine_sv/HG002.annotated/bed/HG002.refine.annotated.delly.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)
#write.table(simplebed, ".../combination_sv/step1_refine_sv/HG002.annotated/bed/HG002.refine.annotated.gridss.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)
#write.table(simplebed, ".../combination_sv/step1_refine_sv/HG002.annotated/bed/HG002.refine.annotated.lumpy.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)
#write.table(simplebed, ".../combination_sv/step1_refine_sv/HG002.annotated/bed/HG002.refine.annotated.manta.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)
#write.table(simplebed, ".../combination_sv/step1_refine_sv/HG002.annotated/bed/HG002.refine.annotated.svaba.bed", quote=FALSE, sep='\t', row.names=FALSE, col.names=FALSE)


